servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/CompletableProcessor.java

```diff
@@ -84,25 +87,32 @@ public void onError(Throwable t) {
         terminate(TerminalNotification.error(t));
     }
 
-    private void terminate(TerminalNotification terminalSignal) {
-        if (terminalSignalUpdater.compareAndSet(this, null, terminalSignal)) {
-            notifyListeners(terminalSignal);
-        }
+    private void terminateLateSubscriber(Subscriber subscriber) {
+        TerminalNotification terminalSignal = this.terminalSignal;
+        assert terminalSignal != null;
+        terminalSignal.terminate(subscriber);
     }
 
-    private void notifyListeners(TerminalNotification terminalSignal) {
-        Throwable delayedCause = null;
-        Subscriber subscriber;
-        while ((subscriber = subscribers.poll()) != null) {
-            try {
-                terminalSignal.terminate(subscriber);
-            } catch (Throwable cause) {
-                delayedCause = catchUnexpected(delayedCause, cause);
+    private void terminate(TerminalNotification terminalSignal) {
+        if (this.terminalSignal == null) {
```

Comment:
Note that `this.terminalSignal` is no longer atomic and the visibility is provided by the `queueUpdater` atomic operations. The `Subscriber` API doesn't allow for concurrent access so this shouldn't violate the API but the value maybe overridden if no external synchronization is provided.

Reply:
Consider clarifying this in the comment in code

Reply:
Actually, it makes me feel a bit uncomfortable that late subscribers may terminate differently. WDYT if instead of a `CLOSED` node inside `ConcurrentStack` we will intoduce a subtype `TerminalNode` which will keep `terminalSignal`? `ConcurrentStack` will have 2 generic types: `ConcurrentStack<Subscriber, TerminalNotification>.

Reply:
yes I can do this. the extra memory barriers adds overhead in multithreaded cases, however it still is an improvement over what we had originally.

```
# non-atomic terminal
CompletableProcessorBenchmark.mtAdd20Complete  thrpt    5    951658.114 ±  15566.060  ops/s

# atomic terminal (with volatile coordination between threads)
CompletableProcessorBenchmark.mtAdd20Complete  thrpt    5    715757.301 ±  22471.221  ops/s 
```

Reply:
addressed in https://github.com/apple/servicetalk/pull/1298